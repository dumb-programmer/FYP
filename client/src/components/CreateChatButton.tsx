import { useRef } from "react";
import CreateChatModal from "./CreateChatModal";

export default function CreateChatButton() {
    const chatModalRef = useRef<HTMLDialogElement>(null);

    return <>
        <button onClick={() => chatModalRef.current?.showModal()}>Create Chat</button>
        <CreateChatModal dialogRef={chatModalRef} />
    </>
}