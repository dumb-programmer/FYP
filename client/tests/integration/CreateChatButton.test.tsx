import React from "react";
import { render, screen } from "@testing-library/react";
import { expect, describe, it, vi, beforeEach, afterEach } from "vitest";
import CreateChatButton from "../../src/components/CreateChatButton";
import userEvent from "@testing-library/user-event";

const renderComponent = () => render(<CreateChatButton />);

describe("CreateChatButton", () => {
    beforeEach(() => {
        HTMLDialogElement.prototype.showModal = vi.fn(function () {
            this.style.display = 'block';
            this.setAttribute('open', '');
        });

        HTMLDialogElement.prototype.close = vi.fn(function () {
            this.style.display = 'none';
            this.removeAttribute('open');
        });
    });

    afterEach(() => {
        vi.clearAllMocks()
    })

    it("renders correctly", () => {
        const { container } = renderComponent();

        expect(container).toMatchSnapshot();
    });

    it("CreateChatModal is closed by default", async () => {
        renderComponent();

        expect(screen.queryByRole("dialog")).not.toBeInTheDocument();
    });

    it("clicking on create chat button opens the dialog", async () => {
        renderComponent();

        const createChatBtn = screen.getByRole("button");

        await userEvent.click(createChatBtn);

        expect(screen.getByRole("dialog")).toBeInTheDocument();
    });
})