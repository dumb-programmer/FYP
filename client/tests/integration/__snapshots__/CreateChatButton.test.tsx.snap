// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CreateChatButton > renders correctly 1`] = `
<div>
  <button
    class="btn ghost"
    data-testid="create-chat-btn"
  >
    <svg
      aria-hidden="true"
      data-slot="icon"
      fill="currentColor"
      height="20"
      viewBox="0 0 16 16"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M8.75 3.75a.75.75 0 0 0-1.5 0v3.5h-3.5a.75.75 0 0 0 0 1.5h3.5v3.5a.75.75 0 0 0 1.5 0v-3.5h3.5a.75.75 0 0 0 0-1.5h-3.5v-3.5Z"
      />
    </svg>
    Create Chat
  </button>
  <dialog
    class="modal"
    data-testid="modal"
    id=":r0:"
  >
    <div
      class="modal-box"
    >
      <h1
        class="text-2xl font-bold"
      >
        Create Chat
      </h1>
      <form
        class="flex flex-col gap-4 mt-4"
        id="create-chat-form"
      >
        <div
          class="form-control"
        >
          <label
            for="name"
          >
            Name
          </label>
          <input
            class="input input-bordered"
            id="name"
            name="name"
            required=""
            type="text"
          />
        </div>
        <div
          class="form-control"
        >
          <label
            for="document"
          >
            Document
          </label>
          <input
            class="input file-input -ml-4 mt-2"
            id="document"
            name="document"
            required=""
            type="file"
          />
        </div>
      </form>
      <div
        class="modal-action"
      >
        <button
          class="btn btn-ghost"
          type="button"
        >
          Cancel
        </button>
        <button
          class="btn btn-primary"
          form="create-chat-form"
          type="submit"
        >
          Create
        </button>
      </div>
    </div>
  </dialog>
</div>
`;
